node/core/pvf/src/worker_common.rs

```diff
@@ -178,16 +177,19 @@ pub async fn tmpfile(prefix: &str) -> io::Result<PathBuf> {
 
 pub fn worker_event_loop<F, Fut>(debug_id: &'static str, socket_path: &str, mut event_loop: F)
 where
-	F: FnMut(UnixStream) -> Fut,
+	F: FnMut(Handle, UnixStream) -> Fut,
 	Fut: futures::Future<Output = io::Result<Never>>,
 {
-	let err = async_std::task::block_on::<_, io::Result<Never>>(async move {
-		let stream = UnixStream::connect(socket_path).await?;
-		let _ = async_std::fs::remove_file(socket_path).await;
-
-		event_loop(stream).await
-	})
-	.unwrap_err(); // it's never `Ok` because it's `Ok(Never)`
+	let rt = Runtime::new().expect("Creates tokio runtime. If this panics the worker will die and the host will detect that and deal with it.");
+	let handle = rt.handle();
+	let err = rt
+		.block_on(async move {
+			let stream = UnixStream::connect(socket_path).await?;
+			let _ = tokio::fs::remove_file(socket_path).await;
+
+			event_loop(handle.clone(), stream).await
+		})
+		.unwrap_err(); // it's never `Ok` because it's `Ok(Never)`
```

Comment:
```suggestion
		// it's never `Ok` because it's `Ok(Never)`
		.unwrap_err(); 
		
		// We don't want tokio to wait for the tasks to finish. We want to bring down the worker 
		// as fast as possible. Otherwise the worker may waits that some malicious parachain validation finishes...
		rt.shutdown_background();
```

Reply:
You should write some test with the `halt` parachain ensuring that the worker shuts down. When you don't have the mentioned changes, it should wait until the validation is finished (what we don't want!)

Reply:
There already are tests using `halt`:

https://github.com/paritytech/polkadot/blob/m-cat%2Freplace-async-std-pvf/node/core/pvf/tests/it/main.rs#L1

I can add some timing information to `terminates_on_timeout`, to make sure that a timed-out job takes > execution_timout, but < the lenient timeout.

Reply:
Hmm good point. I think I know why it terminates, because we don't keep the Tokio handle not alive in the blocking thread as we do it in Substrate. Nevertheless, we should still do these changes for the future.

Reply:
My bad, I totally missed that there were suggested changes. Yes, I think it's worth having this here!